torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -452,73 +477,42 @@ def _sync_param_groups(source: List[Dict[Any, Any]], destination: List[Dict[Any,
-        network requests have been issued. The parameters which are part of a bucket become tensor views.
+    def _setup_flat_buffers(self) -> None:
+        """Make all params which are on the same device and tied to the same rank views of a single buffer.
+        This is used at construction time, and anytime parameter trainability is changed (frozen or unfrozen) and
+        `_update_trainable` is called.
         """
 
-        # Allocate one buffer per rank and per device to group the small parameters
-        for device, per_device in self.per_device_params.items():
-            self.buckets[device] = [
-                torch.zeros(self.bucket_max_size, dtype=per_device[0][0].dtype, device=device)
-                for _ in range(len(per_device))
-            ]
-
-        # Pack the smallest elements in a bucket, depending on their owner shard-wise
         for device, per_rank_params in self.per_device_params.items():
+            # Only wipe the existing buckets if there are none
+            # (could be that this is called twice, when trainability changes)
+            if device not in self.buckets.keys():
+                self.buckets[device] = []
+
+            # Make parameters a view of the bucket
             for dst_rank, params in enumerate(per_rank_params):
-                offset = 0
+                if len(params) > 0:
 
-                for param in params:
-                    # Criteria to decide whether this parameter is to be bucketed or not:
-                    # - enough room in the bucket
-                    if param.requires_grad and (offset + param.numel()) < self.bucket_max_size:
-                        self.should_bucket_param.append(True)
+                    # Clone the non-trainable params, if in a bucket it will get destroyed
+                    for param in filter(lambda x: not x.requires_grad, params):
+                        param.data = param.data.detach().clone()
 
-                        # This parameter becomes a view of the bucket
-                        offset_next = offset + param.numel()
+                    # Merge all the trainable params in a single bucket
+                    trainable_params = list(filter(_is_trainable, params))
+                    buffer_size = sum(map(lambda x: x.numel(), trainable_params))
+                    bucket = torch.empty(buffer_size, dtype=params[0].dtype, device=device)
+                    offset = 0
 
-                        self.buckets[device][dst_rank][offset:offset_next] = param.data.flatten()
-                        param.data = self.buckets[device][dst_rank][offset:offset_next].view_as(param.data)
+                    for param in trainable_params:
+                        offset_next = offset + param.numel()
+                        bucket[offset:offset_next].copy_(param.data.flatten())
+                        param.data = bucket[offset:offset_next].view_as(param.data)
```

Comment:
I wonder if we should still keep the bucket size knob and allow users to choose whether params should be bucket views or not. Because changing `param.data` might break application if user created separate views on parameters. See the following code snippet. Changing `x.data` won't update `y`.

```python
>>> import torch
>>> x = torch.zeros(2, 2)
>>> y = x.view(1, 4)
>>> y
tensor([[0., 0., 0., 0.]])
>>> x
tensor([[0., 0.],
        [0., 0.]])
>>> x.data = torch.ones(2, 2)
>>> x
tensor([[1., 1.],
        [1., 1.]])
>>> y
tensor([[0., 0., 0., 0.]])
```

Does it make sense if we keep the bucket size knob. And then -1 means using bucket view and one bucket per device as implemented in this PR, and other bucket size values means dedicated bucket and no view.

Reply:
yes makes sense, but in that case I would make it a binary choice, no ? If you agree not to touch .data -> ok for tensor view buckets, else don't bucket for safety ? Without the tensor views the buckets are a lot of code for not-so-great effects I think

Reply:
One option might be we keep the PR as-is (always use bucket view) and clarify the side-effect in the documents, so that we don't surprise users. If we see requests to add the no-view option, we can add that later. If this looks OK to you, I will stamp the PR.

Reply:
I think your points where very valid, and adding an option to disable that is not a lot of work, I would just do that. Adding the partial buckets with copies and unroll is not worth it though, in my view, also because it complexifies the code a fair bit and people looking for the best perf all around will probably dig into more complex solutions